In [1]:
## Standard Imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
## Importing the OS and JSON Modules
import os,json

In [3]:
info = pd.read_csv('Data/superhero_info - superhero_info.csv')
powers = pd.read_csv('Data/superhero_powers - superhero_powers.csv')

In [4]:
info.head()

Hero|Publisher Gender               Race Alignment  \
0          A-Bomb|Marvel Comics   Male              Human      good   
1  Abe Sapien|Dark Horse Comics   Male      Icthyo Sapien      good   
2            Abin Sur|DC Comics   Male            Ungaran      good   
3     Abomination|Marvel Comics   Male  Human / Radiation       bad   
4   Absorbing Man|Marvel Comics   Male              Human       bad   

  Hair color Eye color Skin color  \
0    No Hair    yellow    Unknown   
1    No Hair      blue       blue   
2    No Hair      blue        red   
3    No Hair     green    Unknown   
4    No Hair      blue    Unknown   

                                   Measurements  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  
2   {'Height': '185.0 cm', 'Weight': '90.0 kg'}  
3  {'Height': '203.0 cm', 'Weight': '441.0 kg'}  
4  {'Height': '193.0 cm', 'Weight': '122.0 kg'}

In [6]:
info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Hero|Publisher  463 non-null    object
 1   Gender          463 non-null    object
 2   Race            463 non-null    object
 3   Alignment       463 non-null    object
 4   Hair color      463 non-null    object
 5   Eye color       463 non-null    object
 6   Skin color      463 non-null    object
 7   Measurements    463 non-null    object
dtypes: object(8)
memory usage: 29.1+ KB


In [5]:
powers.head()

hero_names                                             Powers
0      3-D Man         Agility,Super Strength,Stamina,Super Speed
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...
3     Abin Sur                                 Lantern Power Ring
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...

In [7]:
powers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 667 entries, 0 to 666
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   hero_names  667 non-null    object
 1   Powers      667 non-null    object
dtypes: object(2)
memory usage: 10.5+ KB


# I. Clean the files and combine them into one final DataFrame.

In [10]:
# splitting Hero|Publisher
info[['Hero', 'Publisher']] = info['Hero|Publisher'].str.split('|', 
                                                               expand=True)
info.head(2)

Hero|Publisher Gender           Race Alignment Hair color  \
0          A-Bomb|Marvel Comics   Male          Human      good    No Hair   
1  Abe Sapien|Dark Horse Comics   Male  Icthyo Sapien      good    No Hair   

  Eye color Skin color                                  Measurements  \
0    yellow    Unknown  {'Height': '203.0 cm', 'Weight': '441.0 kg'}   
1      blue       blue   {'Height': '191.0 cm', 'Weight': '65.0 kg'}   

         Hero          Publisher  
0      A-Bomb      Marvel Comics  
1  Abe Sapien  Dark Horse Comics

In [11]:
# drop Hero|Publisher
info.drop('Hero|Publisher', axis=1, inplace=True)
info.head(2)

Gender           Race Alignment Hair color Eye color Skin color  \
0   Male          Human      good    No Hair    yellow    Unknown   
1   Male  Icthyo Sapien      good    No Hair      blue       blue   

                                   Measurements        Hero          Publisher  
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}      A-Bomb      Marvel Comics  
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  Abe Sapien  Dark Horse Comics

In [15]:
# check for type
measure = info.loc[0,'Measurements']
print(type(measure))
measure

<class 'str'>


"{'Height': '203.0 cm', 'Weight': '441.0 kg'}"

In [16]:
# replace ' with "
info['Measurements'] = info['Measurements'].str.replace("'",'"')
# apply json.loads
info['Measurements'] = info['Measurements'].apply(json.loads)
info['Measurements'].head(2)

0    {'Height': '203.0 cm', 'Weight': '441.0 kg'}
1     {'Height': '191.0 cm', 'Weight': '65.0 kg'}
Name: Measurements, dtype: object

In [18]:
# confirm that dict conversion was successful
measure = info.loc[0,'Measurements']
print(type(measure))
measure

<class 'dict'>


{'Height': '203.0 cm', 'Weight': '441.0 kg'}

In [19]:
# separate the keys into separate cols
height_weight = info['Measurements'].apply(pd.Series)
height_weight

Height    Weight
0    203.0 cm  441.0 kg
1    191.0 cm   65.0 kg
2    185.0 cm   90.0 kg
3    203.0 cm  441.0 kg
4    193.0 cm  122.0 kg
..        ...       ...
458  183.0 cm   83.0 kg
459  165.0 cm   52.0 kg
460   66.0 cm   17.0 kg
461  170.0 cm   57.0 kg
462  185.0 cm   81.0 kg

[463 rows x 2 columns]

In [20]:
# concat to info
info = pd.concat((info, height_weight), axis=1)
info.head(2)

Gender           Race Alignment Hair color Eye color Skin color  \
0   Male          Human      good    No Hair    yellow    Unknown   
1   Male  Icthyo Sapien      good    No Hair      blue       blue   

                                   Measurements        Hero  \
0  {'Height': '203.0 cm', 'Weight': '441.0 kg'}      A-Bomb   
1   {'Height': '191.0 cm', 'Weight': '65.0 kg'}  Abe Sapien   

           Publisher    Height    Weight  
0      Marvel Comics  203.0 cm  441.0 kg  
1  Dark Horse Comics  191.0 cm   65.0 kg

In [21]:
# drop measurments
info.drop('Measurements', axis=1, inplace=True)
info.head(2)

Gender           Race Alignment Hair color Eye color Skin color        Hero  \
0   Male          Human      good    No Hair    yellow    Unknown      A-Bomb   
1   Male  Icthyo Sapien      good    No Hair      blue       blue  Abe Sapien   

           Publisher    Height    Weight  
0      Marvel Comics  203.0 cm  441.0 kg  
1  Dark Horse Comics  191.0 cm   65.0 kg

In [22]:
# dropping str portions of height and weight
info['Height'] = info['Height'].str.replace(' cm', '')
info['Weight'] = info['Weight'].str.replace(' kg', '')
info.head(2)

Gender           Race Alignment Hair color Eye color Skin color        Hero  \
0   Male          Human      good    No Hair    yellow    Unknown      A-Bomb   
1   Male  Icthyo Sapien      good    No Hair      blue       blue  Abe Sapien   

           Publisher Height Weight  
0      Marvel Comics  203.0  441.0  
1  Dark Horse Comics  191.0   65.0

In [30]:
# converting to numeric
info['Height'] = info['Height'].astype(float)
info['Weight'] = info['Weight'].astype(float)

In [31]:
info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 463 entries, 0 to 462
Data columns (total 10 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Gender      463 non-null    object 
 1   Race        463 non-null    object 
 2   Alignment   463 non-null    object 
 3   Hair color  463 non-null    object 
 4   Eye color   463 non-null    object 
 5   Skin color  463 non-null    object 
 6   Hero        463 non-null    object 
 7   Publisher   463 non-null    object 
 8   Height      463 non-null    float64
 9   Weight      463 non-null    float64
dtypes: float64(2), object(8)
memory usage: 36.3+ KB


In [33]:
powers.head(1)

hero_names                                      Powers
0    3-D Man  Agility,Super Strength,Stamina,Super Speed

In [34]:
# check category
power = powers.loc[0,'Powers']
print(type(power))
power

<class 'str'>


'Agility,Super Strength,Stamina,Super Speed'

In [48]:
# split powers str into list
powers['Powers_split'] = powers['Powers'].str.split(',')
powers.head()

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                        Powers_split  
0    [Agility, Super Strength, Stamina, Super Speed]  
1  [Accelerated Healing, Durability, Longevity, S...  
2  [Agility, Accelerated Healing, Cold Resistance...  
3                               [Lantern Power Ring]  
4  [Accelerated Healing, Intelligence, Super Stre...

In [49]:
# check category
power = powers.loc[0,'Powers_split']
print(type(power))
power

<class 'list'>


['Agility', 'Super Strength', 'Stamina', 'Super Speed']

In [50]:
# explode col
exploded = powers.explode('Powers_split')
exploded.head()

hero_names                                             Powers  \
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
0    3-D Man         Agility,Super Strength,Stamina,Super Speed   
1     A-Bomb  Accelerated Healing,Durability,Longevity,Super...   

          Powers_split  
0              Agility  
0       Super Strength  
0              Stamina  
0          Super Speed  
1  Accelerated Healing

In [51]:
# save unique values
cols_to_make = exploded['Powers_split'].dropna().unique()
cols_to_make

array(['Agility', 'Super Strength', 'Stamina', 'Super Speed',
       'Accelerated Healing', 'Durability', 'Longevity', 'Camouflage',
       'Self-Sustenance', 'Cold Resistance', 'Underwater breathing',
       'Marksmanship', 'Weapons Master', 'Intelligence', 'Telepathy',
       'Immortality', 'Reflexes', 'Enhanced Sight', 'Sub-Mariner',
       'Lantern Power Ring', 'Invulnerability', 'Animation',
       'Super Breath', 'Dimensional Awareness', 'Flight', 'Size Changing',
       'Teleportation', 'Magic', 'Dimensional Travel',
       'Molecular Manipulation', 'Energy Manipulation', 'Power Cosmic',
       'Energy Absorption', 'Elemental Transmogrification',
       'Fire Resistance', 'Natural Armor', 'Heat Resistance',
       'Matter Absorption', 'Regeneration', 'Stealth', 'Power Suit',
       'Energy Blasts', 'Energy Beams', 'Heat Generation', 'Danger Sense',
       'Phasing', 'Force Fields', 'Hypnokinesis', 'Invisibility',
       'Enhanced Senses', 'Jump', 'Shapeshifting', 'Elasticity',
 

In [53]:
# make new cols for each power cat (ohe)
for col in cols_to_make:
    powers[col] = powers['Powers_split'].str.contains(col, regex=False)
    # regex=False to return boolean
powers.head()

hero_names                                             Powers  \
0      3-D Man         Agility,Super Strength,Stamina,Super Speed   
1       A-Bomb  Accelerated Healing,Durability,Longevity,Super...   
2   Abe Sapien  Agility,Accelerated Healing,Cold Resistance,Du...   
3     Abin Sur                                 Lantern Power Ring   
4  Abomination  Accelerated Healing,Intelligence,Super Strengt...   

                                        Powers_split  Agility  Super Strength  \
0    [Agility, Super Strength, Stamina, Super Speed]     True            True   
1  [Accelerated Healing, Durability, Longevity, S...    False            True   
2  [Agility, Accelerated Healing, Cold Resistance...     True            True   
3                               [Lantern Power Ring]    False           False   
4  [Accelerated Healing, Intelligence, Super Stre...    False            True   

   Stamina  Super Speed  Accelerated Healing  Durability  Longevity  ...  \
0     True         True                False       False      False  ...   
1     True        False                 True        True       True  ...   
2     True        False                 True        True       True  ...   
3    False        False                False       False      False  ...   
4     True         True                 True       False      False  ...   

   Weather Control  Omnipresent  Omniscient  Hair Manipulation  Nova Force  \
0            False        False       False              False       False   
1            False        False       False              False       False   
2            False        False       False              False       False   
3            False        False       False              False       False   
4            False        False       False              False       False   

   Odin Force  Phoenix Force  Intuitive aptitude  Melting  Changing Armor  
0       False          False               False    False           False  
1       False          False               False    False           False  
2       False          False               False    False           False  
3       False          False               False    False           False  
4       False          False               False    False           False  

[5 rows x 170 columns]

In [54]:
powers.drop(['Powers', 'Powers_split'], axis=1, inplace=True)
powers.head(2)

hero_names  Agility  Super Strength  Stamina  Super Speed  \
0    3-D Man     True            True     True         True   
1     A-Bomb    False            True     True        False   

   Accelerated Healing  Durability  Longevity  Camouflage  Self-Sustenance  \
0                False       False      False       False            False   
1                 True        True       True        True             True   

   ...  Weather Control  Omnipresent  Omniscient  Hair Manipulation  \
0  ...            False        False       False              False   
1  ...            False        False       False              False   

   Nova Force  Odin Force  Phoenix Force  Intuitive aptitude  Melting  \
0       False       False          False               False    False   
1       False       False          False               False    False   

   Changing Armor  
0           False  
1           False  

[2 rows x 168 columns]

In [55]:
# merge dfs
merged = pd.merge(info, powers, left_on='Hero', right_on='hero_names',
                  how='inner')
merged.head(2)

Gender           Race Alignment Hair color Eye color Skin color        Hero  \
0   Male          Human      good    No Hair    yellow    Unknown      A-Bomb   
1   Male  Icthyo Sapien      good    No Hair      blue       blue  Abe Sapien   

           Publisher  Height  Weight  ... Weather Control  Omnipresent  \
0      Marvel Comics   203.0   441.0  ...           False        False   
1  Dark Horse Comics   191.0    65.0  ...           False        False   

   Omniscient  Hair Manipulation  Nova Force  Odin Force  Phoenix Force  \
0       False              False       False       False          False   
1       False              False       False       False          False   

   Intuitive aptitude  Melting  Changing Armor  
0               False    False           False  
1               False    False           False  

[2 rows x 178 columns]

In [57]:
merged.drop('hero_names', axis=1, inplace=True)
merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 463 entries, 0 to 462
Columns: 177 entries, Gender to Changing Armor
dtypes: bool(167), float64(2), object(8)
memory usage: 115.3+ KB


# II. Use your combined DataFrame to answer the following questions.

Compare the average weight of super powers who have Super Speed to those who do not.

In [62]:
super_speed = merged[merged['Super Speed']==True]
super_speed['Super Speed'].value_counts()

True    198
Name: Super Speed, dtype: int64

In [64]:
rest = merged[merged['Super Speed']==False]
rest['Super Speed'].value_counts()

False    265
Name: Super Speed, dtype: int64

In [65]:
super_speed['Weight'].mean()

129.40404040404042

In [66]:
rest['Weight'].mean()

101.77358490566037

Average weight of heroes w/super speed = 129.40 while avg weight of heroes w/out super speed = 101.77.

What is the average height of heroes for each publisher?

In [67]:
merged.groupby('Publisher').mean()

Height      Weight   Agility  Super Strength   Stamina  \
Publisher                                                                       
DC Comics          181.920290  104.188406  0.420290        0.550725  0.485507   
Dark Horse Comics  176.909091  101.818182  0.636364        0.818182  0.636364   
George Lucas       159.600000   77.400000  0.800000        0.000000  0.600000   
Image Comics       211.000000  405.000000  1.000000        1.000000  1.000000   
Marvel Comics      191.545455  119.579125  0.397306        0.609428  0.481481   
Shueisha           171.500000   64.500000  1.000000        1.000000  1.000000   
Star Trek          181.500000   79.000000  0.000000        0.500000  1.000000   
Team Epic TV       180.750000   72.000000  0.500000        0.500000  0.250000   
Unknown            178.000000   83.000000  1.000000        0.000000  1.000000   

                   Super Speed  Accelerated Healing  Durability  Longevity  \
Publisher                                                                    
DC Comics             0.442029             0.253623    0.413043   0.181159   
Dark Horse Comics     0.545455             0.454545    0.727273   0.363636   
George Lucas          0.600000             0.400000    0.200000   0.200000   
Image Comics          1.000000             0.000000    1.000000   0.000000   
Marvel Comics         0.414141             0.309764    0.471380   0.185185   
Shueisha              1.000000             0.750000    0.000000   0.750000   
Star Trek             0.000000             0.000000    0.000000   0.500000   
Team Epic TV          0.000000             0.250000    0.000000   0.250000   
Unknown               0.000000             0.000000    0.000000   0.000000   

                   Camouflage  ...  Weather Control  Omnipresent  Omniscient  \
Publisher                      ...                                             
DC Comics            0.014493  ...         0.007246          0.0         0.0   
Dark Horse Comics    0.181818  ...         0.000000          0.0         0.0   
George Lucas         0.000000  ...         0.000000          0.0         0.0   
Image Comics         0.000000  ...         0.000000          0.0         0.0   
Marvel Comics        0.047138  ...         0.010101          0.0         0.0   
Shueisha             0.250000  ...         0.250000          0.0         0.0   
Star Trek            0.000000  ...         0.000000          0.0         0.0   
Team Epic TV         0.000000  ...         0.000000          0.0         0.0   
Unknown              0.000000  ...         0.000000          0.0         0.0   

                   Hair Manipulation  Nova Force  Odin Force  Phoenix Force  \
Publisher                                                                     
DC Comics                   0.000000    0.000000    0.000000       0.000000   
Dark Horse Comics           0.000000    0.000000    0.000000       0.000000   
George Lucas                0.000000    0.000000    0.000000       0.000000   
Image Comics                0.000000    0.000000    0.000000       0.000000   
Marvel Comics               0.003367    0.006734    0.006734       0.003367   
Shueisha                    0.000000    0.000000    0.000000       0.000000   
Star Trek                   0.000000    0.000000    0.000000       0.000000   
Team Epic TV                0.000000    0.000000    0.000000       0.000000   
Unknown                     0.000000    0.000000    0.000000       0.000000   

                   Intuitive aptitude   Melting  Changing Armor  
Publisher                                                        
DC Comics                         0.0  0.000000            0.00  
Dark Horse Comics                 0.0  0.090909            0.00  
George Lucas                      0.0  0.000000            0.00  
Image Comics                      0.0  0.000000            0.00  
Marvel Comics                     0.0  0.000000            0.00  
Shueisha                          0.0  0.000000            